x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameTransformCheckpoint.java

```diff
@@ -0,0 +1,268 @@
+        this.timestampMillis = in.readLong();
+        this.checkpoint = in.readLong();
+        this.indicesCheckpoints = readCheckpoints(in.readMap());
+        this.timeUpperBoundMillis = in.readLong();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        final String humanReadableSuffix = "_string";
+        builder.startObject();
+
+        // the id, doc_type and checkpoint is only internally used for storage, the user-facing version gets embedded
+        if (params.paramAsBoolean(DataFrameField.FOR_INTERNAL_STORAGE, false)) {
+            builder.field(DataFrameField.ID.getPreferredName(), id);
+            builder.field(CHECKPOINT.getPreferredName(), checkpoint);
+            builder.field(DataFrameField.INDEX_DOC_TYPE.getPreferredName(), NAME);
+        }
+
+        builder.field(TIMESTAMP.getPreferredName(), timestampMillis);
+        if (timeUpperBoundMillis > 0) {
+            builder.field(TIME_UPPER_BOUND.getPreferredName(), timeUpperBoundMillis);
+        }
+
+        if (params.paramAsBoolean("human", false)) {
+            // be human readable on API level
+            builder.field(TIMESTAMP.getPreferredName() + humanReadableSuffix, Instant.ofEpochMilli(timestampMillis).atZone(ZoneOffset.UTC));
+            if (timeUpperBoundMillis > 0) {
+                builder.field(TIME_UPPER_BOUND.getPreferredName() + humanReadableSuffix,
+                        Instant.ofEpochMilli(timeUpperBoundMillis).atZone(ZoneOffset.UTC));
+            }
+        }
+
+        builder.startObject(INDICES.getPreferredName());
+        for (Entry<String, long[]> entry : indicesCheckpoints.entrySet()) {
+            builder.array(entry.getKey(), entry.getValue());
+        }
+        builder.endObject();
+
+        builder.endObject();
+        return builder;
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    public long getTimestamp() {
+        return timestampMillis;
+    }
+
+    public long getCheckpoint() {
+        return checkpoint;
+    }
+
+    public Map<String, long[]> getIndicesCheckpoints() {
+        return indicesCheckpoints;
+    }
+
+    public long getTimeUpperBound() {
+        return timeUpperBoundMillis;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(id);
+        out.writeLong(timestampMillis);
+        out.writeLong(checkpoint);
+        out.writeGenericValue(indicesCheckpoints);
+        out.writeLong(timeUpperBoundMillis);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) {
+            return true;
+        }
+
+        if (other == null || getClass() != other.getClass()) {
+            return false;
+        }
+
+        final DataFrameTransformCheckpoint that = (DataFrameTransformCheckpoint) other;
+
+        return this.timestampMillis == that.timestampMillis && this.checkpoint == that.checkpoint
```

Comment:
Maybe a comment to explain why `id` and `indicesCheckpoints` are not considered here, especially given they are used in `hashCode` which breaks the general contract of `equals` and `hashCode`

Reply:
they are compared in `matches(that)`, the last argument, I will add a comment it isn't easy to spot this